teuthology/repo_utils.py

```diff
@@ -138,6 +142,53 @@ def clone_repo(repo_url, dest_path, branch, shallow=True):
         raise GitError("git clone failed!")
 
 
+def rsstrip(s, suffix):
+    return s[:-len(suffix)] if s.endswith(suffix) else s
+
+
+def lsstrip(s, prefix):
+    return s[len(prefix):] if s.startswith(prefix) else s
+
+
+def remote_ref_from_ref(ref, remote='origin'):
+    if ref.startswith('refs/pull/'):
+        return 'refs/remotes/' + remote + lsstrip(ref, 'refs')
+    elif ref.startswith('refs/heads/'):
+        return 'refs/remotes/' + remote + lsstrip(ref, 'refs/heads')
+    raise GitError("Unsupported ref '%s'" % ref)
+
+
+def local_branch_from_ref(ref):
+    if ref.startswith('refs/pull/'):
+        s = lsstrip(ref, 'refs/pull/')
+        s = rsstrip(s, '/merge')
+        s = rsstrip(s, '/head')
+        return "PR#%s" % s
+    elif ref.startswith('refs/heads/'):
+        return lsstrip(ref, 'refs/heads/')
+    raise GitError("Unsupported ref '%s', try 'refs/heads/' or 'refs/pull/'" % ref)
+
+
+def fetch_refspec(ref):
+    if '/' in ref:
+        remote_ref = remote_ref_from_ref(ref)
+        return "+%s:%s" % (ref, remote_ref)
+    else:
+        # looks like a branch name
+        return ref
+
+
+def clone_repo_ref(repo_url, dest_path, ref):
+    branch_name = local_branch_from_ref(ref)
+    remote_ref = remote_ref_from_ref(ref)
+    misc.sh('git init %s' % dest_path)
+    misc.sh('git remote add origin %s' % repo_url, cwd=dest_path)
+    #misc.sh('git fetch --depth 1 origin %s' % fetch_refspec(ref),
+    #                                                        cwd=dest_path)
+    fetch_branch(dest_path, ref)
+    misc.sh('git checkout -b %s %s' % (branch_name, remote_ref),
+                                                            cwd=dest_path)
+
```

Comment:
one more newline here

Reply:
there is a 2 new lines not three, see cwd=dest_path at the end.